workflow/controller/dag.go

```diff
@@ -55,13 +55,20 @@ func (d *dagContext) taskNodeName(taskName string) string {
 	return fmt.Sprintf("%s.%s", d.boundaryName, taskName)
 }
 
+// nodeTaskName formulates the corresponding task name for a dag node. Note that this is not simply the inverse of
+// taskNodeName. A task name might be from an expanded task, in which case it will not have an explicit task defined for it.
+// When that is the case, we formulate the name of the original expanded task by removing the fields after "("
+func (d *dagContext) taskNameFromNodeName(nodeName string) string {
```

Comment:
After much thinking, I've found this to be the most robust way to get a task name from a node name: expanding all tasks when to match them to the node is costly and would require a code refactor, storing a node name to task name map is futile because tasks are not expanded on every operation of the Workflow, adding a field to NodeStatus is undesirable because it exposes implementation details to the API and the field is only used by DAG Nodes.

This function is safe because node names generated from task names are deterministic, so we should always be able to recover the node name from the task name.

If the naming convention of node names changes, this function will also have to change. To ensure that this happens I have created tests that cover this and have also added assert statements that ensure a node name can be reversed to the correct task name every time one is created.